gtsam/hybrid/HybridGaussianFactorGraph.cpp

```diff
@@ -98,6 +98,12 @@ GaussianMixtureFactor::Sum sumFrontals(
     } else if (f->isContinuous()) {
       deferredFactors.push_back(
           boost::dynamic_pointer_cast<HybridGaussianFactor>(f)->inner());
+
+    } else if (f->isDiscrete()) {
```

Comment:
Yeah there was one test that was failing without this. You can probably comment it out and find out which one.

Reply:
I remember now. If we have discrete only factors it goes to the next condition and errors out since it is not an orphan. That shouldn't happen right? If we only have discrete factors left, we should just be returning a discrete decision tree and not erroring out.

@ProfFan can you please explain what was your rationale here?

Reply:
I think because this is a GaussianMixtureFactor, there should not be any discrete factor added to it

Reply:
This is a Hybrid Factor Graph method so a discrete factor can be present.

Reply:
I mean the return type is `GaussianMixtureFactor::Sum` (`DT<GFG>`), so this function should not process any discrete info?

Reply:
We're iterating over the factors in the graph so it's possible to have discrete factors. In that case we're not processing them, we just move on.

Reply:
@ProfFan can you give me the reason why discrete factors shouldn't be processed here? Without this update, we can't run elimination even on simple factor graphs as shown in the tests.

Reply:
Because we eliminate discrete variables last, if any discrete variable appear here something is wrong

Reply:
If we see a discrete factor, it is just being skipped and not being considered in elimination. So there is no discrete factor in the GaussianMixture.

If we don't have this condition it goes to the next condition which throws an exception which is not correct.

Please see the whole function or better yet, try it out locally to get a better understanding.

Reply:
It says `sumFrontals`, and when we do hybrid the frontals can **only** be continuous variables, so there should not be any discrete factors on it.

Reply:
Why not?

Reply:
And in that case what happens when the factor graph has for example a prior factor on a discrete variable?

Reply:
Fan, I don't think you're completely understanding this change and you don't seem to be reading my responses.
Moreover this is as much your project as mine so rather than provide cursory comments, please a look at this in detail and providing a resolution to the problem.

I have mentioned before that this change is necessary since a hybrid factor graph can have discrete only factors, and this check only serves to skip over discrete factors so we don't error out. Only hybrid or continuous factors are actually processed in this function.

Reply:
@ProfFan it's your original code that is looping over all the factors. Are you saying that when we loop over all the factors, we can't have discrete factors? That doesn't seem right to me.

Reply:
I apologize if my comments look cursory to you. However I think because in the `HybridJunctionTree` special treatment has already been done to make sure that the arguments to `EliminateHybrid` will never have discrete factors. So if you use `eliminatePartialSequential` etc etc there is no problem.

Reply:
The error only happens when you use `EliminateHybrid` directly. When I wrote this I my mental model is that `EliminateHybrid` is like an internal function and will not be called externally, so I did not consider this case in your unit test.

Reply:
I am not opposing to adding this branching. I am a bit worried that people could start to use this function without thinking about the preconditions when this operation is valid. So I prefer to keep this function as-is, or add some comments in the function Doxygen as a warning.

Reply:
Okay now this makes sense. In that case, you could have mentioned this before rather than simply saying this is wrong. I am not aware of your mental model and there aren't any comments about the same.

I believe we should still test `EliminateHybrid` individually. Letting users know what the assumptions are is the right way here, but we should also ensure we handle these kind of edge cases so it makes using the API easier. Can you please comment what would you like me to add as a note in the docstring and I'll add that to this PR?

Reply:
We will need to talk about the actual situation that will happen when this is called on an arbitrary factor graph. What I can think of is
```doxygen
Function to eliminate variables **under the following assumptions**:
1. When the ordering is fully continuous, and the graph only contains continuous and hybrid factors
2. When the ordering is fully discrete, and the graph only contains discrete factors
Any usage outside of this is incorrect.

\warning This function is not meant to be used with arbitrary hybrid factor graphs. For example, if there exists continuous parents, and one tries to eliminate a discrete variable (as specified in the ordering), the result will be INCORRECT and there will be NO error raised.
```

Reply:
How does that sound to you?

Reply:
Sounds good to me.